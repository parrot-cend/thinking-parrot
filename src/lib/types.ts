export interface Dictionary<T = any> {
  [index: string]: T
}

export enum VNodeType {
  Element = 'element',
  Text = 'text',
  Unknown = 'unknown'
}

export interface FormItemSchema {
  model?: string
  label?: string
  tag: string
  prop?: Dictionary<string>
}

export interface ButtonSchema {
  type: string
  size: string
  text: string
}

export interface FormSchema {
  formItem: Array<FormItemSchema>
  buttons: Array<ButtonSchema>
}

export interface OptionsSchema extends Dictionary<FormSchema> {
  queryForm: FormSchema
  // displayForm: FormSchema
  // submitForm: FormSchema
}
