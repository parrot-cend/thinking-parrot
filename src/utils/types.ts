export interface Dictionary<T = any> {
  [index: string]: T
}

export type ContextOptions = Dictionary<string>
