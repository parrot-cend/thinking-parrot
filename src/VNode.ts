import { Dictionary } from '@/utils/types'

class VNode {
  private result: string
  constructor(tag: string, options: Dictionary<string> = {}) {
    const properties = Object.keys(options)
      .map(key => `${key}=\"${options[key]}\"`)
      .join(' ')
    const propertiesStr = properties ? ' ' + properties : ''
    this.result = `<${tag}${propertiesStr}>{slot}</${tag}>`
  }
  public toString(): string {
    return this.result.replace(/\{slot\}/g, '')
  }
  public insert(context: VNode | string): VNode {
    // TODO: 计算缩进层级, 优化代码美观度
    if (context instanceof VNode) {
      this.result = this.result.replace('{slot}', `\n${context.result}\n`)
    } else if (typeof context === 'string') {
      this.result = this.result.replace('{slot}', `\n${context}\n`)
    } else {
      throw new Error('Only can insert string or VNode instance')
    }
    return this
  }
}

export function createElement(tag: string, options?: Dictionary<string>): VNode {
  return new VNode(tag, options)
}
