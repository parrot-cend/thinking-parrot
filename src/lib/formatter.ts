import prettier from 'prettier'

export function formatByPrettier(str: string): string {
  return prettier.format(str, {
    parser: 'html',
    semi: false,
    bracketSpacing: true,
    htmlWhitespaceSensitivity: 'ignore',
    endOfLine: 'lf'
  })
}
