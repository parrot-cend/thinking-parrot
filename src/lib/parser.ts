import { Node, SourceConfig } from './types'
import { createNode, VNode } from './VNode'

type TemplateCode = string

export default function parser(source: SourceConfig): TemplateCode {
  const makeNodeTree = (node: Node.Config): VNode | string => {
    if (typeof node === 'string') {
      return node
    } else {
      const current = createNode(node.tag, node.props)
      if (node.children?.length) {
        current.insertChild(node.children.map(c => makeNodeTree(c)))
      }
      return current
    }
  }
  return Object.values(source)
    .reduce((r, conf) => r.insertChild(conf.map(childConf => makeNodeTree(childConf))), createNode('template'))
    .toString()
}
