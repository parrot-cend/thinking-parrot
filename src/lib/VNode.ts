import { formatByPrettier } from './formatter'
import { Dictionary, Node } from './types'

export interface VNodeOptions {
  props?: Dictionary<Node.Prop>
  children?: Array<VNode>
}

export class VNode {
  public tagName: string = ''
  public props: Array<Node.Prop>
  public children: Array<VNode> = []

  constructor(tag: string, props?: Array<Node.Prop> | Node.Prop) {
    this.tagName = tag
    if (props) {
      this.props = Array.isArray(props) ? props : [props]
    } else {
      this.props = []
    }
    return this
  }
  private makeTagStr(prop?: string, children?: string): string {
    return children?.length ? `<${this.tagName}${prop}>${children}</${this.tagName}>` : `<${this.tagName}${prop} />`
  }
  private makePropStr(): string {
    if (this.props.length > 0) {
      return (
        ' ' +
        this.props
          .map(prop => {
            const realKey = /[a-z][A-Z]/.test(prop.key) ? prop.key.replace(/([A-Z])/g, '-$1').toLowerCase() : prop.key
            if (prop.type === Node.PropType.String) {
              return `${realKey}="${prop.value}"`
            } else if (prop.type === Node.PropType.Boolean) {
              return !prop.value ? realKey : `:${realKey}="${prop.value || false}"`
            } else if (prop.type === Node.PropType.Expression) {
              return `:${realKey}="${prop.value}"`
            } else {
              throw new Error(`
                Invalid Prop Type. Node infomation below:\n
                tagName: ${this.tagName}\n
                props: ${JSON.stringify(this.props)}\n
                errorProp: key: ${JSON.stringify(this.props)}
              `)
            }
          })
          .join(' ')
      )
    } else {
      return ''
    }
  }
  private makeChildrenStr(): string {
    return this.children.reduce<string>((all: string, cur: VNode) => {
      if (typeof cur === 'string') {
        return all + cur
      } else {
        return all + cur.toString()
      }
    }, '')
  }
  public toString(inline?: boolean): string {
    const result = this.makeTagStr(this.makePropStr(), this.makeChildrenStr())
    if (inline) {
      return result
    } else {
      return formatByPrettier(result)
    }
  }
  public insertChild(child: VNode | Array<VNode>): this {
    if (Array.isArray(child)) {
      this.children.push(...child)
    } else {
      this.children.push(child)
    }
    return this
  }
  public removeChild(pos: number): this {
    this.children.splice(pos, 1)
    return this
  }
}

export function createNode(tag: string, props?: Array<Node.Prop> | Node.Prop): VNode {
  return new VNode(tag, props)
}
