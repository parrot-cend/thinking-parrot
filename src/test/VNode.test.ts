import { Node } from '../lib/types'
import { createNode, VNode } from '../lib/VNode'

describe('VNode', () => {
  it('return VNode instance', () => {
    expect(createNode('')).toBeInstanceOf(VNode)
  })
  it('return correct VNode', () => {
    const tagName = 'test'
    const props: Node.Prop = { key: 'test', type: Node.PropType.String, value: '' }
    const node = createNode(tagName, props)
    expect(node.tagName).toEqual(tagName)
    expect(node.props).toEqual([props])
  })
  it('return correct nested string', () => {
    const node = createNode('template')
    node.insertChild(createNode('div'))
    expect(node.toString()).toMatchSnapshot()
  })
  it('remove node correctly', () => {
    const node = createNode('template')
    node.insertChild(createNode('div1'))
    node.insertChild(createNode('div2'))
    node.removeChild(0)
    expect(node.toString()).toMatchSnapshot()
  })
})
