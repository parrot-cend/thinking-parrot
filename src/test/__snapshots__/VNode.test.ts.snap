// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VNode remove node correctly 1`] = `
"<template><div2 /></template>
"
`;

exports[`VNode return correct nested string 1`] = `
"<template><div /></template>
"
`;
